app/experimenter/static/js/design-page.js

```diff
@@ -0,0 +1,494 @@
+
+    this.state = {
+      type: props.expType,
+      pref_key: "",
+      pref_type: "",
+      pref_branch: "",
+      addon_experiment_id: "",
+      addon_release_url: "",
+      design: "",
+      variants: [{
+        ratio: null,
+        name: "",
+        description: "",
+        value: "",
+        is_control: true,
+      }, {
+        ratio: null,
+        name: "",
+        description: "",
+        value: "",
+        is_control: false,
+      }],
+      errors: {}
+    };
+  }
+
+  componentDidMount() {
+    const that = this;
+
+    let url;
+    if (this.state.type == "pref") {
+      url = `/api/v1/experiments/${this.slug}/design-pref`
+    } else if (this.state.type == "addon") {
+      url = `/api/v1/experiments/${this.slug}/design-addon`
+    } else {
+      url = `/api/v1/experiments/${this.slug}/design-generic`
+    }
+
+    fetch(url)
+      .then(function(response) {
+        return response.json();
+      })
+      .then(function(json) {
+        if (json.variants == "") {
+          return that.setState({
+            type: json.type,
+            pref_key: json.pref_key,
+            pref_type: json.pref_type,
+            pref_branch: json.pref_branch,
+            addon_experiment_id: json.addon_experiment_id,
+            addon_release_url:json.addon_release_url,
+            design: json.design
+          })
+        } else {
+          return that.setState({
+            type: json.type,
+            pref_key: json.pref_key,
+            pref_type: json.pref_type,
+            pref_branch: json.pref_branch,
+            addon_experiment_id: json.addon_experiment_id,
+            addon_release_url:json.addon_release_url,
+            design: json.design,
+            variants: json.variants
+          });
+        }
+      });
+  }
+
+  addBranch(e) {
+    e.preventDefault();
+    this.setState({
+      variants: this.state.variants.concat({
+        ratio: null,
+        name: "",
+        description: "",
+        value: "",
+        is_control: false,
+      })
+    });
+  }
+
+  removeBranch(e) {
+    e.preventDefault();
+    this.state.variants.splice(e.target.dataset.index, 1);
+
+    this.setState({ variants: this.state.variants });
+  }
+
+  updateRatio(e) {
+    this.state.variants[e.target.dataset.index].ratio = e.target.value
+    this.setState({variants: this.state.variants})
+  }
+
+  updateName(e) {
+    this.state.variants[e.target.dataset.index].name = e.target.value
+    this.setState({variants: this.state.variants})
+
+  }
+
+  updateDescription(e) {
+    var stateCopy = { ...this.state.variants };
+    stateCopy[e.target.dataset.index].description = e.target.value;
+
+    this.setState(stateCopy);
+  }
+
+  updateValue(e) {
+    var stateCopy = { ...this.state.variants };
+    stateCopy[e.target.dataset.index].value = e.target.value;
+
+    this.setState(stateCopy);
+  }
+
+  handleInputChange(e) {
+    if (e.target.name == "pref-name") {
```

Comment:
Alternatively could you create an object to relate these fields to their paired state key names and then just to the look up and set once?

Reply:
huhmmm?